### PR TITLE
Cert-Manager (wip)

### PR DESCRIPTION
**Pull request by gambol99**
- updating with migration instructions and providing a reference in the certificates docs

### CODE REVIEW

File: how-to-docs/cert-manager.md [hunk 1]
```diff
@@ -0,0 +1,235 @@
+## **Cert Manager**
+----
+
+The ACP platform presently has two certificate management services, the first contender was [kube-cert-manager](https://github.com/PalmStoneGames/kube-cert-manager). The project due to a forever changing landscape gradually became deprecated and has now been replaced by the [cert-manager](https://github.com/jetstack/cert-manager). Note ACP will continue to support the kube-cert-manager and the internal cfssl service while they are still in use, we do however recommend shifting over to the cert-manager as aside from security fix we won't be performing anymore updates on these services.
+
+Without wishing to duplicate documentation, which are can get from both the [readme](https://github.com/jetstack/cert-manager/blob/master/README.md) or it's official [documentation](https://cert-manager.readthedocs.io/en/latest/), the cert-manager can effectively replace two services
```
Comment (marcinc) [inline how-to-docs/cert-manager.md]: maybe `which can be found in both the [readme] or its official [documentation]`

File: how-to-docs/cert-manager.md [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+
+- kube-cert-manager: used to acquire certificates from LetsEncrypt.
+- cfssl: an internal Cloudflare service used to generate internal certificate _(usaually to encrypt between ingress and pod)_.
+
+### **How-tos**
+
+#### **As a developer I already have a certificate from the legacy kube-cert-manger, can I migrate?**
+
+Migrating from the former [kube-cert-manager](https://github.com/PalmStoneGames/kube-cert-manager) over to [cert-manager](https://github.com/jetstack/cert-manager) means creating the certificate request as below and removing the annotations from the ingress. However, the safety way would be to
```
Comment (marcinc) [inline how-to-docs/cert-manager.md]: `the safe way would be to`

File: how-to-docs/cert-manager.md [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+
+- Create a new Certificate resource and point to a **new** secret name _(thus keeping the old one incase)_.
+- Push out the change and wait for the certificate to be furfilled.
```
Comment (marcinc) [inline how-to-docs/cert-manager.md]: `fulfilled`

File: how-to-docs/cert-manager.md [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+- Once you have the certificate you can update your ingress to use the new secret,**remove** the annotations and use the Certificate resource there after.
```
Comment (marcinc) [inline how-to-docs/cert-manager.md]: `thereafter`

File: how-to-docs/cert-manager.md [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+    kind: Issuer
+  commonName: site.svc.project.cluster.local
+  dnsNames:
+  - localhost
+  - 127.0.0.1
+```
+
+#### **As a developer I want to retrieve a certificate for my external service**
+
+Lets assume we have an externally facing site which we wish to expose via ingress and we want a valid Letsencrypt certificate. Two have two things to note here;
```
Comment (marcinc) [inline how-to-docs/cert-manager.md]: `Two have` -> `We have`?

File: how-to-docs/cert-manager.md [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+$ kubectl -n project get challenge
+```
+
+**Network Policies**
+
+In order to handle the http01 challenge the requestor is provided an ephermal token which must be handled back to LetsEncrypt on the path `http://domain/.well-known/acme-challenge/`, thus validating to them you own the domain's your requesting for. Cert-manager handles the http01 resolver by;
+
+- creating a order resource within the namespace.
+- a controller pick up the order and creates a challenge resource from it, creating an ephermal pod, service and ingress in your namespace and routing the `/.well-known/acme-challenge/`.
+- the controller checks the path/s has been provision via the ingress beforehand and validates the order, the certificate controller is then free to inform letsencrypt to can call us back.
```
Comment (marcinc) [inline how-to-docs/cert-manager.md]: `to can call us back` -> `to call us back`

File: how-to-docs/cert-manager.md [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+- it continues to probe the request and once validate pulls down the certificate.
```
Comment (marcinc) [inline how-to-docs/cert-manager.md]: validate -> validated

File: how-to-docs/cert-manager.md [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+
+**IMPORTANT** All this requires the user's to add a network policy to permit the callback, as ACP by default deny's all traffic.
```
Comment (marcinc) [inline how-to-docs/cert-manager.md]: deny's -> denies

File: how-to-docs/cert-manager.md (hunk 5 above)
Comment (KashifSaadat) [reply how-to-docs/cert-manager.md]: Two have or not two have!

File: how-to-docs/cert-manager.md [hunk 9]
```diff
@@ -0,0 +1,235 @@
+## **Cert Manager**
+----
+
+The ACP platform presently has two certificate management services, the first contender was [kube-cert-manager](https://github.com/PalmStoneGames/kube-cert-manager). The project due to a forever changing landscape gradually became deprecated and has now been replaced by the [cert-manager](https://github.com/jetstack/cert-manager). Note ACP will continue to support the kube-cert-manager and the internal cfssl service while they are still in use, we do however recommend shifting over to the cert-manager as aside from security fix we won't be performing anymore updates on these services.
```
Comment (oyelekci) [inline how-to-docs/cert-manager.md]: I'd suggest:
- splitting the second sentence after comma to improve readability. 
- make 'security fix' plural, and add a comma after it.

File: how-to-docs/cert-manager.md [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+  kind: Ingress
+  metadata:
+    annotations:
+      # @NOTE: we choose http challenge
+      certmanager.k8s.io/acme-challenge-type: http01
+      # @NOTE: this will enable the cert-manager to handle this resource
+      certmanager.k8s.io/enabled: "true"
+      ingress.kubernetes.io/affinity: cookie
+      ingress.kubernetes.io/force-ssl-redirect: "true"
+      ingress.kubernetes.io/secure-backends: "true"
```
Comment (jaykeshur) [inline how-to-docs/cert-manager.md]: remove secure-backends and replace with `ingress.kubernetes.io/backend-protocol: "HTTPS"`

File: how-to-docs/cert-manager.md [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+      kubernetes.io/ingress.class: nginx-external
+    name: example
+  spec:
+    rules:
+    - host: www.example.com
+      http:
+        paths:
+        - backend:
+            serviceName: service_name
+            servicePort: 443
```
Comment (jaykeshur) [inline how-to-docs/cert-manager.md]: Change this to 10443, just to be more consistant on our platform

File: how-to-docs/certificates.md [hunk 12]
```diff
@@ -12,8 +12,11 @@ You would want all communication between the user, through to the service and se
 
 #### **Certificates**
 
+ACP currently supports two modes for certificates, the first one legacy and being deprecated.
+
 - [kube-cert-manager and cfssl](https://github.com/UKHomeOffice/application-container-platform/blob/master/how-to-docs/kube-cert-manager.md)
+- [cert-mamager](https://github.com/UKHomeOffice/application-container-platform/blob/master/how-to-docs/cert-manager.md)
```
Comment (jaykeshur) [inline how-to-docs/certificates.md]: correct - mamager

**Pull request merged**